src/swap/useSwapQuote.ts

```diff
@@ -1,27 +1,238 @@
+import { FetchQuoteResponse, Field, ParsedSwapAmount, SwapTransaction } from 'src/swap/types'
+import { TokenBalance, TokenBalanceWithAddress } from 'src/tokens/slice'
 import Logger from 'src/utils/Logger'
+import { publicClient } from 'src/viem'
+import { estimateFeesPerGas } from 'src/viem/estimateFeesPerGas'
 import networkConfig from 'src/web3/networkConfig'
 import { walletAddressSelector } from 'src/web3/selectors'
+import { Address, EstimateGasExecutionError, Hex, encodeFunctionData } from 'viem'
 
-interface ExchangeRate {
+// Apply a multiplier for the decreased swap amount to account for the
+// varying gas costs of different swap providers (or even the same swap)
+const DECREASED_SWAP_AMOUNT_GAS_COST_MULTIPLIER = 1.2
+
+export interface QuoteResult {
   toTokenAddress: string
   fromTokenAddress: string
   swapAmount: BigNumber
   price: string
   provider: string
   estimatedPriceImpact: BigNumber | null
+  preparedTransactions?: PreparedTransactionsResult
+}
+
+interface PreparedTransactionsPossible {
+  type: 'possible'
+  approveTransaction: TransactionRequestCIP42 & { amountToApprove: bigint }
+  swapTransaction: TransactionRequestCIP42
+}
+
+export interface PreparedTransactionsNeedDecreaseSwapAmountForGas {
+  type: 'need-decrease-swap-amount-for-gas'
+  maxGasCost: BigNumber
+  feeCurrency: TokenBalance
+  decreasedSwapAmount: BigNumber
+}
+
+export interface PreparedTransactionsNotEnoughBalanceForGas {
+  type: 'not-enough-balance-for-gas'
+  feeCurrencies: TokenBalance[]
+}
+
+export type PreparedTransactionsResult =
+  | PreparedTransactionsPossible
+  | PreparedTransactionsNeedDecreaseSwapAmountForGas
+  | PreparedTransactionsNotEnoughBalanceForGas
+
+function createBaseApproveTransaction(
+  updatedField: Field,
+  unvalidatedSwapTransaction: SwapTransaction
+) {
+  const { guaranteedPrice, sellTokenAddress, buyAmount, sellAmount, allowanceTarget, from } =
+    unvalidatedSwapTransaction
+  const amountType: string =
+    updatedField === Field.TO ? ('buyAmount' as const) : ('sellAmount' as const)
+
+  const amountToApprove =
+    amountType === 'buyAmount'
+      ? BigInt(new BigNumber(buyAmount).times(guaranteedPrice).toFixed(0, 0))
+      : BigInt(sellAmount)
+
+  const data = encodeFunctionData({
+    abi: erc20.abi,
+    functionName: 'approve',
+    args: [allowanceTarget as Address, amountToApprove],
+  })
+
+  const approveTx: TransactionRequestCIP42 & { amountToApprove: bigint } = {
+    from: from as Address,
+    to: sellTokenAddress as Address,
+    data,
+    type: 'cip42',
+    amountToApprove,
+  }
+
+  return approveTx
+}
+
+function createBaseSwapTransaction(unvalidatedSwapTransaction: SwapTransaction) {
+  const { from, to, value, data, gas } = unvalidatedSwapTransaction
+
+  const swapTx: TransactionRequestCIP42 & { gas: bigint } = {
+    from: from as Address,
+    to: to as Address,
+    value: BigInt(value ? value : 0),
+    data: data as Hex,
+    gas: BigInt(gas), // This isn't entirely accurate
```

Comment:
> This isn't entirely accurate

what does that mean?

Reply:
It's related to this: https://www.notion.so/valora-inc/Fee-currency-selection-logic-4c207244893748bd85e23b754334f42d?pvs=4#8b7c27d31ebf4fca981f81e9411f86ee

I can add the link as a comment.

Reply:
thanks